
# MIDAS Database

MIDAS databases can be downloaded in a on-demand manner, and therefore pre-downloading the entire database is not necessary. At this moment, we host MIDAS DB for two public databases (UHGG and GTDB) on the public S3 bucket (`s3://microbiome-igg`). MIDAS DB can be sliced efficiently based on the species needed for the analysis. Therefore, for quick start, or users with limited disk size, we recommend taking advantage of this feature. We also provide command to download the MIDAS DB, with the default assignment of the representative genomes. For advanced usage of reassigning representative genomes for the SNPs calling analysis, access to AWS S3 bucket is required. The default MIDAS DB can be downloaded at link. We strongly advice against downloading the whole bucket via `aws s3 cp` command, due to the large amounts of intermediate files created during generating the MIDAS DB.

## List Available Databases

```
$ midas2 database --list
uhgg 286997 genomes from 4644 species version 1.0
gtdb 286997 genomes from 4644 species version r202
```

## Download 

- Download the MIDAS DB for UHGG to local directory: `midasdb_uhgg` = /path/to/local/midas/db

```
$ python -m midas2 database --download --midasdb_name uhgg --midasdb_dir $midasdb_uhgg
```

- Download the MIDAS DB for UHGG to local directory: `midasdb_gtdb` = /path/to/local/midas/db

```
$ python -m midas2 database --download --midasdb_name gtdb --midasdb_dir $midasdb_gtdb
```

# MIDAS 2.0

After installing MIDAS 2.0, users can test the main functionality of MIDAS 2.0 with the provided unit test data (`tests`). 

```
cd MIDAS2.0
bash tests/run_midas.sh 8
```


We also provided the scripts to build custom databases for any given collection of geomes. 

```
cd MIDAS2.0
bash tests/build_midasdb.sh 8
```