
Install Conda Environment

<!--
TEST TEST TEST
-->

```
conda env create --name=midas2.0 --quiet --file midas2.yml
cpanm Bio::SearchIO::hmmer --force # Temporary fix for Prokka
```

Install MIDAS 2.0
```
pip install midas2
```

Download Test Data
```
mkdir -p midas2_quick_start
cd midas2_quick_start

wget reads.tar.gz # TODO
tar -zxf reads.tar.gz
```

Run MIDAS 2.0 SNV Module

```
sample_name="sample1"
midas2 run_species --sample_name ${sample_name} -1 reads/${sample_name}_R1.fastq.gz \
    --midasdb_name uhgg --midasdb_dir my_midasdb_uhgg --num_cores 8 my_midas2_output

midas2 run_snps --sample_name ${sample_name} -1 reads/${sample_name}_R1.fastq.gz \
    --midasdb_name uhgg --midasdb_dir my_midasdb_uhgg --chunk_size 500000 \
    --select_by median_marker_coverage,unique_fraction_covered --select_threshold=5,0.5 \
    --num_cores 8 my_midas2_output
```

Now, you get your first MIDAS 2.0 single-sample SNV module result. One of the expected output is `my_midas2_output/sample1/snps/snps_summary.tsv`, which summarize the read mapping and pileup results of the species in the restricted species profile.


For advanced users, a more complex unit test for all main features of MIDAS 2.0:

```
git clone https://github.com/czbiohub/MIDAS2.0.git .
cd MIDAS2.0

bash tests/midas2_analysis.sh 8
```
