# Metagenomic Intra-Species Diversity Analysis 2.0

Metagenomic Intra-Species Diversity Analysis ([MIDAS](https://genome.cshlp.org/content/26/11/1612)) is an integrated pipeline for profiling strain-level genomic variations in shotgun metagenomic data. The standard MIDAS workflow harnesses a reference database of 5,926 species extracted from 30,000 genomes(MIDAS DB v1.2).

MIDAS 2.0 used the same analysis workflow as the original [MIDAS tool](https://github.com/snayfach/MIDAS), and is engineered to work with more comprehensive MIDAS Reference Databases (MIDAS DBs), and to run on  collections of thousands of samples in a fast and scalable manner. This page is focused specifically on the analysis steps, and for information about the MIDAS DB, including the target layout of MIDAS DB, how to build a new database, users should refer to  MIDASDB can refer to [this page](https://github.com/czbiohub/MIDAS2.0/wiki/4.-MIDAS-2.0-Database). 

***

# MIDAS 2.0 Results Layout

MIDAS 2.0 contains two strain-level analysis modules: population SNVs analysis (SNPs module) and pan-genome CNVs analysis (Genes module).  Each module includes two sequential steps: single-sample analysis and across-samples analysis. 


![Figure 1: MIDAS 2.0 Analysis Modules](https://github.com/czbiohub/MIDAS2.0/blob/master/docs/images/Fig.Modules.png?raw=true "Title")


## Single-sample Results Layout

All single-sample workflow takes as a parameter the path to MIDAS results root directory (`midas_outdir`); and together with `sample_name` constitute the unique output directory `midas_outdir/sample_name`.  All subsequent analysis steps operate within that directory.

Single-sample analysis start with identifying the list of abundantly present species by profiling 15 universal single copy marker genes (subcommand `run_species`), followed by subcommand `run_snps` for single-sample reads pileup and subcommand `run_genes` for single-sample pan-gene content profiling.  

Here is an example of the results layout  of all three single-sample analysis in the local filesystem.

```
Output                                       Producer          Meaning
--------------------------------------------------------------------------------------------------------
{midas_output}/{sample_name}
  |- species
     |- species_profile.tsv                  run_species       Summary of species coverage
     |- markers_profile.tsv                  run_species       Per species marker coverage
  |- snps
     |- snps_summary.tsv                     run_snps          Summary of reads mapping to rep-genome
     |- {species}.snps.tsv.lz4               run_snps          Per species pileups
  |- genes 
     |- genes_summary.tsv                    run_genes         Summary of reads mapping to pan-genome
     |- {species}.genes.tsv.lz4              run_genes         Per species pan-gene coverage
 
 |- temp                        
     |- snps
        |- repgenomes.bam                    run_snps          Rep-genome alignment file
        |- {species}/snps_XX.tsv.lz4
     |- genes
        |- pangenome.bam                     run_genes         Pan-genome alignment file
        |- {species}/genes_XX.tsv.lz4
  |- bt2_indexes                                                  
     |- snps/repgenomes.*                    run_snps          Sample-specific rep-genome database
     |- genes/pangenomes.*                   run_genes         Sample-specific pan-genome database
```

## Across-samples Results Layout

For a collection of samples, population SNVs and pan-genome CNVs can be estimated using subcommands `merge_snps` and `merge_genes`. User need to provide the output directory (`midas_outdir`) as the input parameter, which is the root of the layout below.

```
Output                                           Producer        Meaning
---------------------------------------------------------------------------------------------------------------
species
  |- species_prevalence.tsv                      merge_species   Per species summary statistics across samples
  |- species/species_read_counts.tsv             merge_species   Species-by-sample read counts matrix
  |- species/species_coverage.tsv                merge_species   Species-by-sample marker coverage matrix
  |- species/species_rel_abundance.tsv           merge_species   Species-by-sample relative abundance matrix
snps
  |- snps_summary.tsv                            merge_snps      Alignment summary statistics per sample
  |- {species}/{species}.snps_info.tsv.lz4       merge_snps      Per species metadata for genomic sites
  |- {species}/{species}.snps_freqs.tsv.lz4      merge_snps      Per species site-by-sample MAF matrix
  |- {species}/{species}.snps_depth.tsv.lz4      merge_snps      Per species site-by-sample read depth matrix
genes
  |- genes_summary.tsv                           merge_genes     Alignment summary statistics per sample
  |- {species}/{species}.genes_presabs.tsv.lz4   merge_genes     Per species gene-by-sample pre-abs matrix
  |- {species}/{species}.genes_copynum.tsv.lz4   merge_genes     Per species gene-by-sample copy number matrix
  |- {species}/{species}.genes_depth.tsv.lz4     merge_genes     Per species gene-by-sample read depth matrix
```


***


# Single-sample Analysis

Reference-based variants calling and copy number estimation requires choosing a reference genome(s) as the template. Microbiome data usually contains hundreds of species in one sample, and only species with enough reads coverage (horizontal and vertical) can be used for reliable strain-level analysis. A good reference database should be both representative and comprehensive in terms of the abundant species in the given sample. Therefore, single-sample analysis of MIDAS 2.0 start with building a sample-specific reference database of species selected via profiling 15 universal single copy marker genes (Species module).


## Common Command Line Arguments

All three single-sample analysis subcommands of MIDAS 2.0 share the following command line arguments.

```
positional arguments:
  midas_outdir          Path to base directory to store results. 

optional arguments:
  --sample_name SAMPLE_NAME   
                        Unique sample identifier. The output results will be stored at 
                        midas_outdir/sample_name.  
  -1 R1                 FASTA/FASTQ file containing 1st mate if using paired-end 
                        reads. Otherwise FASTA/FASTQ containing unpaired reads.
  -2 R2                 FASTA/FASTQ file containing 2nd mate if using paired-end reads.
  
  --midasdb_name {uhgg,gtdb,testdb}
                        MIDAS Database name. If not specificed, default option is uhgg.
  --midasdb_dir         Path to Local MIDASDB. If not specified, MIDASDB would be 
                        on-demand downloaded to current directory (.).

  --num_cores INT       Number of physical cores to use (8)
  -h, --help            Show help message and exit.
  -f, --force           Force rebuild of pre-existing outputs.
  -g, --debug           Debug mode: skip cleanup on error, extra prints.
```

