# Metagenomic Intra-Species Diversity Analysis 2.0

Metagenomic Intra-Species Diversity Analysis ([MIDAS](https://genome.cshlp.org/content/26/11/1612)) is an integrated pipeline for profiling strain-level genomic variations in shotgun metagenomic data. The standard MIDAS workflow harnesses a reference database of 5,926 species extracted from 30,000 genomes (MIDAS DB v1.2).

MIDAS 2.0 used the same analysis workflow as the original [MIDAS tool](https://github.com/snayfach/MIDAS), and is engineered to work with more comprehensive MIDAS Reference Databases (MIDAS DBs), and to run on  collections of thousands of samples in a fast and scalable manner. This page is focused specifically on the analysis steps, and for information about the MIDAS DB, including the target layout of MIDAS DB, how to build a new database, users should refer to  MIDASDB can refer to [this page](https://github.com/czbiohub/MIDAS2.0/wiki/4.-MIDAS-2.0-Database). 


## MIDAS 2.0 Workflow

MIDAS 2.0 contains two strain-level reads-to-table analysis modules: population SNVs analysis (SNV module) and pan-genome CNVs analysis (CNV module).  Each module includes two sequential steps: single-sample analysis and across-samples analysis. 


![Figure 1: MIDAS 2.0 Analysis Modules](https://github.com/czbiohub/MIDAS2.0/blob/master/docs/images/Fig.Modules.png?raw=true "Title")


## MIDAS 2.0 Results Layout

### Results Directory

MIDAS 2.0 writes its outputs to a user-specified root directory, which is always passed as a mandatory argument to each of the MIDAS 2.0 command. In this documentation we refer to this directory as:
- `midas_outdir=/path/to/results/root/directory` 

Together with the unique sample name:
- `sample_name=/unique/id/of/sample` 

`${midas_outdir}/${sample_name}` constitute the unique output directory for single-sample analysis. All subsequent analysis steps operate within this directory. For the across-samples SNV or CNV analysis, all subsequent analysis steps operate within the directory `${midas_outdir}`.


### Single-sample Results Layout

MIDAS 2.0 analysis usually starts with the "database customization" which selects sufficiently abundant species in each sample (subcommand `run_species`). After completing this step, users can run either of two strain-level analysis: `run_snps` for single-sample read pileup (SNV module) or `run_genes for pan-gene profiling (CNV module).  Here is an example of the results layout of all three single-sample commands in the local filesystem.

```
Output                                       Producer          Meaning
--------------------------------------------------------------------------------------------------------
{midas_output}/{sample_name}
  |- species
     |- species_profile.tsv                  run_species       Summary of species coverage
     |- markers_profile.tsv                  run_species       Per species marker coverage
  |- snps
     |- snps_summary.tsv                     run_snps          Summary of read mapping to rep-genome
     |- {species}.snps.tsv.lz4               run_snps          Per species pileups
  |- genes 
     |- genes_summary.tsv                    run_genes         Summary of read mapping to pan-genome
     |- {species}.genes.tsv.lz4              run_genes         Per species pan-gene coverage
 
 |- temp                        
     |- snps
        |- repgenomes.bam                    run_snps          Rep-genome alignment file
        |- {species}/snps_XX.tsv.lz4
     |- genes
        |- pangenome.bam                     run_genes         Pan-genome alignment file
        |- {species}/genes_XX.tsv.lz4
  |- bt2_indexes                                                  
     |- snps/repgenomes.*                    run_snps          Sample-specific rep-genome database
     |- genes/pangenomes.*                   run_genes         Sample-specific pan-genome database
```

### Across-samples Results Layout

For a collection of samples, population SNVs and pan-genome CNVs can be estimated using subcommands `merge_snps` and `merge_genes`. 

```
Output                                             Producer        Meaning
---------------------------------------------------------------------------------------------------------------
{midas_output}
  |- species
    |- species_prevalence.tsv                      merge_species   Per species summary statistics across samples
    |- species/species_read_counts.tsv             merge_species   Species-by-sample read counts matrix
    |- species/species_coverage.tsv                merge_species   Species-by-sample marker coverage matrix
    |- species/species_rel_abundance.tsv           merge_species   Species-by-sample relative abundance matrix
  |- snps
    |- snps_summary.tsv                            merge_snps      Alignment summary statistics per sample
    |- {species}/{species}.snps_info.tsv.lz4       merge_snps      Per species metadata for genomic sites
    |- {species}/{species}.snps_freqs.tsv.lz4      merge_snps      Per species site-by-sample MAF matrix
    |- {species}/{species}.snps_depth.tsv.lz4      merge_snps      Per species site-by-sample read depth matrix
  |-genes
    |- genes_summary.tsv                           merge_genes     Alignment summary statistics per sample
    |- {species}/{species}.genes_presabs.tsv.lz4   merge_genes     Per species gene-by-sample pre-abs matrix
    |- {species}/{species}.genes_copynum.tsv.lz4   merge_genes     Per species gene-by-sample copy number matrix
    |- {species}/{species}.genes_depth.tsv.lz4     merge_genes     Per species gene-by-sample read depth matrix
```


