# Metagenomic Intra-Species Diversity Analysis 2.0

Metagenomic Intra-Species Diversity Analysis ([MIDAS](https://genome.cshlp.org/content/26/11/1612)) is an integrated pipeline for profiling strain-level genomic variation and gene copy number variations (CNVs) for metagenomics data. Its analysis steps are run against a database of 5,926 bacterial species extracted from 30,000 genomes (MIDAS DB v1.2).

MIDAS 2.0 represent a reimplementation of the same analysis steps as the original [MIDAS tool](https://github.com/snayfach/MIDAS), but able to operate on the more comprehensive MIDAS Databases (DBs)for a larger collection of samples in a fast and scalable manner. Similar to the original MIDAS, MIDAS 2.0 also presuppose a database construction step has already taken place. This page is focused specifically on the analysis steps, and database construction and download steps can refer [here](https://github.com/czbiohub/MIDAS2.0/wiki/4.-MIDAS-2.0-Database). 

***

# MIDAS 2.0 Results Layout

MIDAS 2.0 contains two strain-level analysis modules: population SNVs analysis (SNPs module) and pan-genome CNVs analysis (Genes module).  Each module includes two sequential steps: single-sample analysis and across-samples analysis. 

[insert a figure here.]

## Single-sample Results Layout

All single-sample workflow takes as a parameter the path to MIDAS results root directory (`midas_outdir`); and together with `sample_name` constitute the unique output directory `output_dir`:=`midas_outdir/sample_name`.  All subsequent analysis steps operate within that directory.

Single-sample analysis start with identifying the list of abundantly present species by profiling 15 universal single copy marker genes (subcommand `run_species`), followed by subcommand `run_snps` for single-sample reads pileup and subcommand `run_genes` for single-sample pan-gene content profiling.  

Here is an example of layout of the results of all three single-sample analysis in the local filesystem.

```
Output                                       Producer             Meaning
--------------------------------------------------------------------------------------------------------
{midas_output}/{sample_name}
  |- species
     |- species_profile.tsv                  midas_run_species    Summary of species coverage
     |- markers_profile.tsv                  midas_run_species    Per species marker coverage
  |- snps
     |- snps_summary.tsv                     midas_run_snps       Summary of reads mapping to rep-genome
     |- {species}.snps.tsv.lz4               midas_run_snps       Per species pileups
  |- genes 
     |- genes_summary.tsv                    midas_run_genes      Summary of reads mapping to pan-genome
     |- {species}.genes.tsv.lz4              midas_run_genes      Per species pan-gene coverage
 
 |- temp                        
     |- snps
        |- repgenomes.bam                                         Rep-genome alignment files
        |- {species}/snps_XX.tsv.lz4
     |- genes
        |- pangenome.bam                                          Pan-genome alignment file
        |- {species}/genes_XX.tsv.lz4
  |- bt2_indexes                                                  
     |- snps/repgenomes.*                                         Sample-specific rep-genome database
     |- genes/pangenomes.*                                        Sample-specific pan-genome database
```

## Across-samples Results Layout

For a collection of samples, population SNVs and pan-genome CNVs can be estimated using subcommands `merge_snps` and `merge_genes`. User need to provide the output directory (`{midas_outdir}`) as the input parameter, which is the root of the layout below.

```
Output                                           Producer        Meaning
---------------------------------------------------------------------------------------------------------------
species
  |- species_prevalence.tsv                      merge_species   Per species summary statistics across samples
  |- species/species_read_counts.tsv             merge_species   Species-by-sample read counts matrix
  |- species/species_coverage.tsv                merge_species   Species-by-sample marker coverage matrix
  |- species/species_rel_abundance.tsv           merge_species   Species-by-sample relative abundance matrix
snps
  |- snps_summary.tsv                            merge_snps      Alignment summary statistics per sample
  |- {species}/{species}.snps_info.tsv.lz4       merge_snps      Per species metadata for genomic sites
  |- {species}/{species}.snps_freqs.tsv.lz4      merge_snps      Per species site-by-sample MAF matrix
  |- {species}/{species}.snps_depth.tsv.lz4      merge_snps      Per species site-by-sample read depth matrix
genes
  |- genes_summary.tsv                           merge_genes     Alignment summary statistics per sample
  |- {species}/{species}.genes_presabs.tsv.lz4   merge_genes     Per species gene-by-sample pre-abs matrix
  |- {species}/{species}.genes_copynum.tsv.lz4   merge_genes     Per species gene-by-sample copy number matrix
  |- {species}/{species}.genes_depth.tsv.lz4     merge_genes     Per species gene-by-sample read depth matrix
```


***

# Single-sample Analysis

Reads mapping based variants calling and copy number estimation requires choosing a reference genome(s) as the template. Microbiome data usually contains hundreds of species in one sample, and only species with enough reads coverage (horizontal and vertical) can be used for reliable strain-level analysis. A good reference database should be both representative and comprehensive in terms of the abundant species in the given sample. Therefore, single-sample analysis of MIDAS 2.0 start with building a sample-specific reference database of species selected via profiling 15 universal single copy marker genes (Species module).

## Abundant Species Detection

The goal of the single-sample Species module is to quickly detect the presence of abundant species in the given sample, to better serve the needs of building proper sample-specific reference databases.  Metagenomic sequencing reads were mapped to 15 universal single copy genes via `hs-blastn`, and uniquely mapped read counts per marker gene were computed. And ambiguous mapped reads were then probabilistically assigned. Only species with more than two marker genes covered with more than two reads are reported. We recommend using `median_marker_coverage` and `unique_fraction_covered` as the metrics to identify the panel of abundant species.  Due to the simplicity of the 15 marker genes, we recommend against using the Species module for taxonomic profiling. 

### Sample command

  ```
  midas2 run_species --sample_name ${sample_name} -1 /path/to/R1 -2 /path/to/R2 \
         --midasdb_name uhgg --midasdb_dir /path/to/local/midasdb --num_cores 8 ${midas_outdir}
  ```

### Output files

- `species_profile.tsv`: sorted in decreasing order of `median_marker_coverage`. 

   ```
   species_id  marker_read_counts  median_marker_coverage  marker_coverage  marker_relative_abundance   unique_fraction_covered
   102470      329                 3.53                    4.16             0.54                        1.00    
   100039      199                 3.45                    2.45             0.32                        0.73
   ```

Refer to [original MIDAS's estimate species abundance](https://github.com/snayfach/MIDAS/blob/master/docs/species.md) for more details.


## Single Nucleotide Variants (SNVs) calling

Depending on the purpose of the experiment, users can decide the list of species upon which sample-specific representative genome (rep-genome) database would be built, and to be genotyped.  For example, with the goal to genotype low abundant species, we suggested to set `median_marker_coverage > 0`, instead of the default `median_marker_coverage > 2`.  Reads were aligned to the sample-specific rep-genome database using Bowtie2, and per genomic site read pileup and nucleotide variation for **all** the species in the rep-genome database are reported by MIDAS 2.0.  MIDSA 2.0 purposely hold any species selection until across-samples SNVs analysis. That being said, per species reads mapping summary are reported in `snps_summay.tsv`, and pileup and variants calling results are organized by species; therefore users can easily select species accordingly based on the `horizontal_coverage` and `vertical_coverage`. 

### Sample command

- Single-sample SNVs calling for all the species in the restricted species profile: `median_marker_coverage` > 2 and `unique_fraction_covered` > 0.5
   
   ```
   midas2 run_species --sample_name ${sample_name} -1 ${R1} -2 ${R2} \
         --midasdb_name uhgg --midasdb_dir /path/to/local/midasdb \
         --num_cores 8 ${midas_outdir}

   midas2 run_snps --sample_name ${sample_name} -1 ${R1} -2 ${R2} \
         --midasdb_name uhgg --midasdb_dir /path/to/local/midasdb \
         --num_cores 12 --fragment_length 2000 \
         --select_by median_marker_coverage,unique_fraction_covered \
         --select_threshold=2,0.5 \
          ${midas_outdir}
    ```

- Single-sample SNVs calling with prebuilt rep-genome database: no need to run the Species module to decide the list of abundant species anymore. 

  - `--select_threshold=-1`: perform reads pileup for all the species in the user-provided Bowtie2 databases.

   ```
   midas2 run_snps --sample_name ${sample_name} -1 ${R1} -2 ${R2} \
          --midasdb_name uhgg --midasdb_dir /path/to/local/midasdb \
          --prebuilt_bowtie2_indexes ${bt2_indexes/${bt2_name} \
          --prebuilt_bowtie2_species ${bt2_indexes/${bt2_name}.species \
          --select_threshold=-1 \
          --num_cores 12 ${midas_outdir}
   ``` 

### Output files

- Reads mapping and pileup summary: `snps_summary.tsv`

   ```
   species_id  genome_length  covered_bases  total_depth  aligned_reads  mapped_reads  fraction_covered mean_coverage
   102470      5774847        4379800        23295116     213486         202365        0.758            5.319
   100039      2852528        1764680        25765375     225417         224630        0.619            14.601
   ```

- Per-species reads pileup results: `{species}.snps.tsv.lz4`

   ```
   ref_id                    ref_pos   ref_allele  depth  count_a  count_c  count_g  count_t
   gnl|Prokka|UHGG143484_2   531422    C           5       0        5       0        0  
   gnl|Prokka|UHGG143484_2   531423    T           6       2        0       0        4  
   gnl|Prokka|UHGG143484_2   531424    A           6       6        0       0        0  
   ```

  - In advanced mode (`--advanced`), single-sample SNPs module would report the variants callings for **all** the genomic sites. We recommend setting `--ignore_ambiguous` to avoid calling variants for site with read counts ties (e.g. gnl|Prokka|UHGG143484_2).

   ```
   ref_id                    ref_pos   ref_allele  depth  count_a  count_c  count_g  count_t  major_allele  minor_allele  major_allele_freq  minor_allele_freq  allele_counts
   gnl|Prokka|UHGG143484_2   531422    C           5       0        5       0        0        C              C            1.000              1.000              1
   gnl|Prokka|UHGG143484_2   531423    T           6       2        0       0        4        T              A            0.667              0.333              2
   gnl|Prokka|UHGG143484_2   531424    A           10      5        0       5        0        A              A            0.500              0.500              1
   ```

Refer to [MIDAS's call single nucleotide polymorphisms](https://github.com/snayfach/MIDAS/blob/master/docs/cnvs.md) for more details.


### New Features

- **Chunkified Pileup**

  Single-sample Pileup was parallelized on the unit of chunk of sites, which is indexed by <species_id, chunk_id>. Only when all chunks from the same species finished processing, chunk-level pileup results would be merged into species-level pileup file (`{species}.snps.tsv.lz4`).


- **Re-assign Representative Genome**

  Users can re-assign representative genome for species of interests by modifying the table of content `genomes.tsv` of MIDAS DB accordingly.


## Pangenome Profiling

Species-level pangenome refers to the set of non-redundant genes (centroids) clustered from all the genomes within one species cluster. Species in the restricted species profile would be concatenated and used to build the sample-specific pangenome database, to which reads were aligned using Bowtie2. Per species per centroid **copy number** were computed in three steps: (1) Per centroid, reads alignment metrics, .e.g `mapped reads` and `mean coverage`, were computed; (2) Per species, median reads coverage of all the mapped centroids corresponding to the 15 universal SCGs; (3) Per centroid, `copy number` were computed and gene presence/absence were inferred.

### Sample command

   ```
   midas2 run_species --sample_name ${sample_name} -1 ${R1} -2 ${R2} \
         --midasdb_name uhgg --midasdb_dir /path/to/local/midasdb \
         --num_cores 8 ${midas_outdir}

   midas2 run_genes --sample_name ${sample_name} -1 ${R1} -2 ${R2} \
          --midasdb_name uhgg --midasdb_dir /path/to/local/midasdb \
          --select_by median_marker_coverage,unique_fraction_covered \
          --select_threshold=2,0.5 \
          --num_cores 12 ${midas_outdir}
   ```

### Output files

- Reads mapping and gene coverage summary: `genes_summary.tsv`: reads aligned to more than one genomic location equally well (`MAPQ=0, 1`) are ignored by default. 

   ```
   species_id  pangenome_size  covered_genes  fraction_covered  mean_coverage  aligned_reads  mapped_reads   marker_coverage
   102470      87136           6122           0.070             9.519          455045         269656         3.447
   100039      79571           2350           0.030             9.462          206684         113644         0.000
   ```

- Per-species gene content profiling results: `{species_id}.genes.tsv.lz4`: 

   ```
   gene_id              gene_length  aligned_reads  mapped_reads  mean_coverage  fraction_covered  copy_number
   UHGG000947_00080     3138         53             6             2.208723       0.102294          0.640716
   UHGG000947_00081     609          27             25            5.083770       0.940887          1.474722
   ```

Refer to [MIDAS's predict pan-genome gene content](https://github.com/snayfach/MIDAS/blob/master/docs/cnvs.md) for more details.

***

# Across-samples Analysis

Population metagenomic SNVs calling and pangenome CNVs estimation can be accomplished by merging single-sample analysis results using the subcommands `merge_snps` and `merge_genes`.  All across-samples subcommands require the input file `samples_list`: a TSV file with `sample_name` and single-sample midas output directory `midas_outdir`. For example, given the following `samples_list`, `merge_snps` would expect to locate `/mnt/cz/hmp-test/SRS011134/snps/snps_summary.tsv`, generate by `run_snps`.

   ```
   sample_name   midas_outdir
   SRS011134     /mnt/cz/hmp-test
   SRS011271     /mnt/cz/hmp-test
   ```


## Population SNVs Calling

1. **<species, sub-samples-lists> selection**

   As mentioned in the single-sample step, MIDAS 2.0 purposely hold any filter upon the single-sample pileup summary until the across-samples step.  The across-samples SNPs analysis restricts attention to "sufficiently well" covered species in "sufficiently many" samples. To be specific, only <species, sample> pair with more than 40% horizontal genome coverage (`fraction_covered`) and 5X vertical genome coverage (`mean_coverage`) would be kept. Furthermore, only "sufficiently prevalent" species with "sufficiently many" (`sample_counts`) would be included for the population SNVs analysis. Therefore, different species may have different lists of relevant samples.


2. **<site, relevant sample> selection**

   For each genomic site, a sample is considered to be "relevant" if the corresponding site read depth falls between the range defined by the input parameters `site_depth` and `site_ratio * genome_depth`; otherwise ignored for the pooled-SNPs compute.  Therefore, different genomic sites from the same species may have different panels of "relevant samples".  And genomic site prevalence can be computed as the ratio of the number of relevant samples for the given site over the total number of relevant samples for the given species.

3. **Relevant site** 

   For each species, a site is considered to be "relevant" if the site prevalence meets the range defined by the input argument `snv_type` and `site_prev`. 


4. **Population SNVs**

   (1) For each **relevant** genomic site, MIDAS 2.0 determines the set of alleles present across **all** relevant samples. Specifically, for each allele (A, C, G, T), `merge_snps` subcommand (1) counts the number of **relevant samples** containing corresponding allele (`scA:scT`), and (2) sums up the corresponding allele's site depths across those samples (`rc_G:rc_T`).

   ```
   site_id               rc_A  rc_C  rc_G  rc_T  sc_A  sc_C  sc_G  sc_T
   UHGG143484_1|810|C    0     10    0     2     0     2     0     2
   UHGG143484_1|905|T    0     10    0     86    0     3     0     2
   ```

   (2) Across-samples major and minor allele for a single site can be computed by the metric specified via the `snps_pooled_method`: sample counts as in `prevalence` and read counts as in `abundance`. For example, the population major allele of site UHGG143484_1|905|T in the above example is T defined by accumulated read counts and C defined by accumulated sample counts. And the population minor allele refers to the second most prevalent/abundant allele.  

   (3) The sample by site matrix of the corresponding site depths and allele frequency of the above calculated across-samples minor allele for each sample will be collected and reported.

   More details about the compute can be found at [Cross-Sample SNP Analysis Tools (xsnp)](https://github.com/czbiohub/xsnp/wiki/Data-Schema-And-Computation)


### Sample commands

   ```
   midas2 merge_snps --samples_list /path/to/sample/lists --num_cores 32 ${merged_midas_outdir}
   ```

### Output files

- Pooled single-sample pileup summary: `snps_summary.tsv`

   ```
   sample_name species_id  genome_length  covered_bases  total_depth  aligned_reads  mapped_reads  fraction_covered  mean_coverage
   sample1     102470      5774847        4379800        23295116     213486         202365        0.758             5.319
   sample1     100039      2852528        1764680        25765375     225497         224630        0.619             14.601
   ```

- Per species SNPs information file: `{species_id}.snps_info.tsv.lz4`.  It contains the computed population major and minor alleles, together with all the metadata.

   ```
   site_id                          major_allele  minor_allele  sample_counts  snp_type  rc_A  rc_C  rc_G  rc_T  sc_A  sc_C  sc_G  sc_T  locus_type  gene_id           site_type  amino_acids
   gnl|Prokka|UHGG143484_1|810|C    C             T             2              bi        0     10     0     2     0     2    0      2    CDS         UHGG143484_00002  4D         G,G,G,G
   gnl|Prokka|UHGG143484_1|905|T    C             T             2              bi        0     2      0     16    0     2    0      2    CDS         UHGG143484_00002  1D         Y,S,C,F
   ```

- Per species site by sample allele frequency matrix of population minor allele: `{species_id}.snps_freq.tsv.lz4`.

  ```
  site_id                           sample1   sample3
  gnl|Prokka|UHGG143484_1|810|C	    0.167     0.167
  gnl|Prokka|UHGG143484_1|905|T     0.889     0.889
  ```

- Per species site by sample site depth matrix: `{species_id}.snps_depth.tsv.lz4`. **Only** accounts for the alleles matching either population major or/and population minor allele.

  ```
  site_id                           sample1  sample3
  gnl|Prokka|UHGG143484_1|810|C	    6        6
  gnl|Prokka|UHGG143484_1|905|T	    9	     9
  ```

Refer to [MIDAS's merge SNPs](https://github.com/snayfach/MIDAS/blob/master/docs/merge_snvs.md) for more details.


## Population CNVs Profiling

Across-samples pan-genome gene content profiling is a straightforward task. MIDAS 2.0 `merge_gene` subcommands merge single-sample gene content profiling results across all the samples listed in the `samples_list`, and further on quantify each gene's presence/absence by compare the `copy_number` with the user defined minimal gene copy number to call a gene present (`min_copy`).


### Sample command

   ```
   midas2 merge_genes --samples_list /path/to/tsv --num_cores 8 ${merged_midas_outdir}
   ```

### Target output files

- Pooled single-sample pileup summary: `genes_summary.tsv`

   ```
   sample_name  species_id  pangenome_size  covered_genes  fraction_covered  mean_coverage  aligned_reads  mapped_reads  marker_depth
   sample1      100247      87136           6122           0.070             9.519          455045         269656        3.447
   sample3      100247      57785           1889           0.033             4.849          108041         30072         0.000
   ```

- Per species gene by sample copy number matrix: `{species_id}.genes_copynum.tsv.lz4`

  ```
  gene_id            sample1     sample3
  UHGG000947_00080   0.640716    0.640716
  UHGG000947_00081   1.474722    1.474722
  ```

- Per species gene by sample presence absence matrix: `{species_id}.genes_preabs.tsv.lz4`
 
  ```
  gene_id             sample1     sample3
  UHGG000947_00080    1           1
  UHGG000947_00081    1           1
  ```

- Per species gene by sample gene average coverage matrix: `{species_id}.genes_depth.tsv.lz4`

  ```
  gene_id             sample1     sample3
  UHGG000947_00080    2.208723    2.208723
  UHGG000947_00081    5.083770    5.083770
  ```


Refer to [MIDAS's merge gene content](https://github.com/snayfach/MIDAS/blob/master/docs/merge_cnvs.md) for more details.


## Across-samples Species Abundance Profile

MIDAS 2.0 also provide the subcommand `merge_species` to merge all the single-sample single copy marker genes profiling results for all the listed samples. This subcommand can be useful when users want to build one Bowtie2 genome database for a panel of samples with similar species abundance profile. For example, by providing the generated `species_prevalence.tsv` and sample selection criteria (`select_by` and `select_threshold`), `build_bowtie2db` subcommand can produce the rep-genome and pan-genome database for the prevalent species meeting the specified filters.

### Sample command

   ```
   midas2 merge_species --samples_list /path/to/sample/lists ${merged_midas_outdir}
   ```

### Output files

- Pooled single-sample marker genes reads mapping summary: `species_prevalence.tsv`

   ```
   species_id  median_abundance  mean_abundance  median_coverage  mean_coverage  sample_counts
   102470      0.462             0.334           3.533            2.355          2
   100039      0.317             0.320           3.455            3.455          3
   ```

- Species by sample median marker genes coverage matrix: `species_marker_median_coverage.tsv`

   ```
   species_id   sample1   sample2    sample3  
   102470       3.533     0.000      3.533
   100039       3.455     3.455      3.455
   ```

- Species by sample average marker genes coverage matrix: `species_marker_coverage.tsv`

   ```
   species_id   sample1   sample2    sample3  
   102470       4.161     0.000      4.199
   100039       2.446     2.446      2.408
   ```

- Species by sample marker genes read counts matrix: `species_marker_read_counts.tsv`

   ```
   species_id   sample1   sample2   sample3  
   102470       329       0.000     332
   100039       199       199       196
   ```

- Species by sample marker genes relative abundance matrix: `species_relative_abundance.tsv`

   ```
   species_id   sample1   sample2    sample3  
   102470       0.539     0.000      0.462
   100039       0.317     0.378      0.265
   ```

Refer to [MIDAS's merge species abundance](https://github.com/snayfach/MIDAS/blob/master/docs/merge_species.md) for more details.
