# Metagenomic Intra-Species Diversity Analysis 2.0

Metagenomic Intra-Species Diversity Analysis ([MIDAS](https://genome.cshlp.org/content/26/11/1612)) is an integrated pipeline for profiling strain-level genomic variations in shotgun metagenomic data. The standard MIDAS workflow harnesses a reference database of 5,926 species extracted from 30,000 genomes(MIDAS DB v1.2).

MIDAS 2.0 represent a reimplementation of the same analysis steps as the original [MIDAS tool](https://github.com/snayfach/MIDAS), but able to operate on the more comprehensive MIDAS Reference Databases (MIDASDBs)for a larger collection of samples in a fast and scalable manner. Similar to the original MIDAS, MIDAS 2.0 also presuppose a database construction step has already taken place. This page is focused specifically on the analysis steps, and MIDASDB can refer [here](https://github.com/czbiohub/MIDAS2.0/wiki/4.-MIDAS-2.0-Database). 

***

# MIDAS 2.0 Results Layout

MIDAS 2.0 contains two strain-level analysis modules: population SNVs analysis (SNPs module) and pan-genome CNVs analysis (Genes module).  Each module includes two sequential steps: single-sample analysis and across-samples analysis. 


![Figure 1: MIDAS 2.0 Analysis Modules](https://github.com/czbiohub/MIDAS2.0/blob/master/docs/images/Fig.Modules.png?raw=true "Title")


## Single-sample Results Layout

All single-sample workflow takes as a parameter the path to MIDAS results root directory (`midas_outdir`); and together with `sample_name` constitute the unique output directory `midas_outdir/sample_name`.  All subsequent analysis steps operate within that directory.

Single-sample analysis start with identifying the list of abundantly present species by profiling 15 universal single copy marker genes (subcommand `run_species`), followed by subcommand `run_snps` for single-sample reads pileup and subcommand `run_genes` for single-sample pan-gene content profiling.  

Here is an example of the results layout  of all three single-sample analysis in the local filesystem.

```
Output                                       Producer          Meaning
--------------------------------------------------------------------------------------------------------
{midas_output}/{sample_name}
  |- species
     |- species_profile.tsv                  run_species       Summary of species coverage
     |- markers_profile.tsv                  run_species       Per species marker coverage
  |- snps
     |- snps_summary.tsv                     run_snps          Summary of reads mapping to rep-genome
     |- {species}.snps.tsv.lz4               run_snps          Per species pileups
  |- genes 
     |- genes_summary.tsv                    run_genes         Summary of reads mapping to pan-genome
     |- {species}.genes.tsv.lz4              run_genes         Per species pan-gene coverage
 
 |- temp                        
     |- snps
        |- repgenomes.bam                    run_snps          Rep-genome alignment file
        |- {species}/snps_XX.tsv.lz4
     |- genes
        |- pangenome.bam                     run_genes         Pan-genome alignment file
        |- {species}/genes_XX.tsv.lz4
  |- bt2_indexes                                                  
     |- snps/repgenomes.*                    run_snps          Sample-specific rep-genome database
     |- genes/pangenomes.*                   run_genes         Sample-specific pan-genome database
```

## Across-samples Results Layout

For a collection of samples, population SNVs and pan-genome CNVs can be estimated using subcommands `merge_snps` and `merge_genes`. User need to provide the output directory (`midas_outdir`) as the input parameter, which is the root of the layout below.

```
Output                                           Producer        Meaning
---------------------------------------------------------------------------------------------------------------
species
  |- species_prevalence.tsv                      merge_species   Per species summary statistics across samples
  |- species/species_read_counts.tsv             merge_species   Species-by-sample read counts matrix
  |- species/species_coverage.tsv                merge_species   Species-by-sample marker coverage matrix
  |- species/species_rel_abundance.tsv           merge_species   Species-by-sample relative abundance matrix
snps
  |- snps_summary.tsv                            merge_snps      Alignment summary statistics per sample
  |- {species}/{species}.snps_info.tsv.lz4       merge_snps      Per species metadata for genomic sites
  |- {species}/{species}.snps_freqs.tsv.lz4      merge_snps      Per species site-by-sample MAF matrix
  |- {species}/{species}.snps_depth.tsv.lz4      merge_snps      Per species site-by-sample read depth matrix
genes
  |- genes_summary.tsv                           merge_genes     Alignment summary statistics per sample
  |- {species}/{species}.genes_presabs.tsv.lz4   merge_genes     Per species gene-by-sample pre-abs matrix
  |- {species}/{species}.genes_copynum.tsv.lz4   merge_genes     Per species gene-by-sample copy number matrix
  |- {species}/{species}.genes_depth.tsv.lz4     merge_genes     Per species gene-by-sample read depth matrix
```


***


# Single-sample Analysis

Reference-based variants calling and copy number estimation requires choosing a reference genome(s) as the template. Microbiome data usually contains hundreds of species in one sample, and only species with enough reads coverage (horizontal and vertical) can be used for reliable strain-level analysis. A good reference database should be both representative and comprehensive in terms of the abundant species in the given sample. Therefore, single-sample analysis of MIDAS 2.0 start with building a sample-specific reference database of species selected via profiling 15 universal single copy marker genes (Species module).


## Common Command Line Arguments

All three single-sample analysis subcommands of MIDAS 2.0 share the following command line arguments.

```
positional arguments:
  midas_outdir          Path to base directory to store results. 

optional arguments:
  --sample_name SAMPLE_NAME   
                        Unique sample identifier. The output results will be stored at 
                        midas_outdir/sample_name.  
  -1 R1                 FASTA/FASTQ file containing 1st mate if using paired-end 
                        reads. Otherwise FASTA/FASTQ containing unpaired reads.
  -2 R2                 FASTA/FASTQ file containing 2nd mate if using paired-end reads.
  
  --midasdb_name {uhgg,gtdb,testdb}
                        MIDAS Database name. If not specificed, default option is uhgg.
  --midasdb_dir         Path to Local MIDASDB. If not specified, MIDASDB would be 
                        on-demand downloaded to current directory (.).

  --num_cores INT       Number of physical cores to use (8)
  -h, --help            Show help message and exit.
  -f, --force           Force rebuild of pre-existing outputs.
  -g, --debug           Debug mode: skip cleanup on error, extra prints.
```

## Abundant Species Detection

The goal of the single-sample Species module is to quickly detect the presence of abundant species in the given sample, to better serve the needs of building proper sample-specific reference databases.  Metagenomic sequencing reads were mapped to 15 universal single copy genes (SCGs) via `hs-blastn`, and uniquely mapped read counts per marker gene were computed. And ambiguous mapped reads were then probabilistically assigned. Only species with more than two marker genes covered with more than two reads are reported in the `species_profile.tsv`. Due to the simplicity of only 15 universal SCG markers, we don't recommend using `run_species` for taxonomic profiling. 

### Sample command

  ```
  midas2 run_species --sample_name ${sample_name} -1 /path/to/R1 -2 /path/to/R2 \
         --midasdb_name uhgg --midasdb_dir /path/to/local/midasdb --num_cores 8 /path/to/base/midas/output
  ```

### Output files

- `species_profile.tsv`: sorted in decreasing order of `median_marker_coverage`. 

   ```
   species_id  marker_read_counts  median_marker_coverage  marker_coverage  marker_relative_abundance   unique_fraction_covered
   102337      4110                28.48                   28.91            0.30                        1.00
   102506      734                 4.98                    4.98             0.05                        0.93
   ```

### Select Species for Strain-level Analysis

The panel of species eligible for strain-level genomic variation analysis is determined by the SCG profiling results. Sample-specific rep-genome and pan-genome database would be built only for the species passing the filter.  We recommend using the combination of `median_marker_coverage` and `unique_fraction_covered` as the metrics to determine the list of abundant species. However, if the goal of the experiment is to genotype low abundant species, then users need to set the parameters properly, e.g. `median_marker_coverage > 0`.  The two subcommands `run_snps` and `run_genes` share the same command line arguments for filtering species:

```
  --select_by SELECT_BY
                        Comma separated columns from species_profile to filter
                        species.
  --select_threshold CHAR
                        Comma separated cutoff correspond to select_by to filter
                        species (> XX) (2, )
```
  
An alternative way to select the species for downstream analysis is to directly provide the list of species. The species in the provided species list is still subject to the `select_threshold` restriction. Users can set `--select_threshold=-1` to escape any filters.

```
  --species_list CHAR   Comma separated list of species ids
```


## Single Nucleotide Variants (SNVs) Calling

Sample-specific rep-genome database of the species in the restricted species profile were built, to which reads were aligned using Bowtie2. Per genomic site read pileup and nucleotide variation for **all** the species in the rep-genome database are reported by MIDAS 2.0.  MIDAS 2.0 purposely hold any species selection based on the pileup results until across-samples SNVs analysis. That being said, per species reads mapping summary are reported in `snps_summay.tsv`, and pileup and variants calling results are organized by species; therefore users can easily select species accordingly based on the reported `horizontal_coverage` and `vertical_coverage` in the `snps_summary.tsv`.

### Sample command

- Single-sample SNVs calling for all the species in the restricted species profile: `median_marker_coverage` > 2 and `unique_fraction_covered` > 0.5
   
   ```
   midas2 run_species --sample_name ${sample_name} -1 /path/to/R1 -2 /path/to/R2\
         --midasdb_name uhgg --midasdb_dir /path/to/local/midasdb \
         --num_cores 8 /path/to/base/midas/output

   midas2 run_snps --sample_name ${sample_name} -1 /path/to/R1 -2 /path/to/R2 \
         --midasdb_name uhgg --midasdb_dir /path/to/local/midasdb \
         --select_by median_marker_coverage,unique_fraction_covered \
         --select_threshold=2,0.5 \
         --num_cores 12 /path/to/base/midas/output
    ```

- Single-sample SNVs calling with prebuilt rep-genome database. Species module is no longer needed.

  - `--select_threshold=-1`: escape species selection filters based on the SCG species profiling. Pilup would be performed for all the species in the user-provided Bowtie2 databases.

   ```
   midas2 run_snps --sample_name ${sample_name} -1 /path/to/R1 -2 /path/to/R2 \
          --midasdb_name uhgg --midasdb_dir /path/to/local/midasdb \
          --prebuilt_bowtie2_indexes /path/to/prebuilt/bowtie2_index \
          --prebuilt_bowtie2_species /path/to/list/of/species/in/bowtie2_index \
          --select_threshold=-1 \
          --num_cores 12 /path/to/base/midas/output
   ``` 

### Output files

- Reads mapping and pileup summary: `snps_summary.tsv`

   ```
   species_id  genome_length  covered_bases  total_depth  aligned_reads  mapped_reads  fraction_covered   mean_coverage
   102506      5339468        2373275        8045342      468667         224553        0.444              3.390
   102337      2749621        2566404        47723458     1479479        1010530       0.933              18.595
   ```

- Per-species reads pileup results: `{species}.snps.tsv.lz4`

   ```
   ref_id                    ref_pos   ref_allele  depth   count_a  count_c  count_g  count_t
   gnl|Prokka|UHGG144544_1   881435    T           11      0        0        0        11
   gnl|Prokka|UHGG144544_1   881436    T           13      0        5        0        8
   gnl|Prokka|UHGG144544_1   881437    T           12      0        6        0        6
   ```

- In the advanced mode (`--advanced`), the per-species pileup results would also report the called variants for **all** the genomic sites. We recommend setting `--ignore_ambiguous` to avoid falsely calling major/minor allele for site with read counts ties (e.g. ref_position: 881437).

   ```
   ref_id                    ref_pos   ref_allele  depth   count_a  count_c  count_g  count_t  major_allele  minor_allele  major_allele_freq  minor_allele_freq  allele_counts
   gnl|Prokka|UHGG144544_1   881435    T           11      0        0        0        11       T             T             1.000              0.000              1
   gnl|Prokka|UHGG144544_1   881436    T           13      0        5        0        8        T             C             0.615              0.385              2
   gnl|Prokka|UHGG144544_1   881437    T           12      0        6        0        6        C             T             0.500              0.500              2
   ```

## Pangenome Profiling

Species-level pangenome refers to the set of non-redundant genes (centroids) clustered from all the genomes within one species cluster. Species in the restricted species profile would be concatenated and used to build the sample-specific pangenome database, to which reads were aligned using Bowtie2. Per species per centroid **copy number** were computed in three steps: (1) Per centroid, reads alignment metrics, .e.g `mapped reads` and `mean coverage`, were computed; (2) Per species, median reads coverage of all the mapped centroids corresponding to the 15 universal single-copy genes (SCGs); (3) Per centroid, `copy number` were computed and gene presence/absence were inferred.

### Sample command

   ```
   midas2 run_species --sample_name ${sample_name} -1 /path/to/R1 -2 /path/to/R2 \
         --midasdb_name uhgg --midasdb_dir /path/to/local/midasdb \
         --num_cores 8 /path/to/base/midas/output

   midas2 run_genes --sample_name ${sample_name} -1 /path/to/R1 -2 /path/to/R2 \
          --midasdb_name uhgg --midasdb_dir /path/to/local/midasdb \
          --select_by median_marker_coverage,unique_fraction_covered \
          --select_threshold=2,0.5 \
          --num_cores 12 /path/to/base/midas/output
   ```

### Output files

- Reads mapping and gene coverage summary: `genes_summary.tsv`. 

  - By default, reads aligned to more than one genomic location equally well (`MAPQ=0, 1`) are ignored.

   ```
   species_id  pangenome_size  covered_genes  fraction_covered  mean_coverage  aligned_reads  mapped_reads   marker_coverage
   102337      15578           4468           0.287             16.213         1650361        450353         20.213
   102506      731186          4733           0.006             3.803          681335         37272          2.140
   ```

- Per-species gene content profiling results: `{species_id}.genes.tsv.lz4`: 

   ```
   gene_id              gene_length  aligned_reads  mapped_reads  mean_coverage  fraction_covered  copy_number
   UHGG143901_00483     555          14             6             2.961538       0.234234          1.384035
   UHGG143901_03589     384          103            57            32.840708      0.294271          15.347667
   UHGG143902_04031     207          9              2             1.737500       0.386473          0.811997
   ```

***

# Across-samples Analysis

Population metagenomic SNVs calling and pangenome CNVs estimation can be accomplished by merging single-sample analysis results using the subcommands `merge_snps` and `merge_genes`.  All across-samples subcommands require the input file `samples_list`: a TSV file with `sample_name` and single-sample midas output directory `midas_outdir`. For example, given the following `samples_list`, `merge_snps` would expect to locate `/home/ubuntu/hmp_mock/midas2_output_uhgg/single_sample/SRR172903/snps/snps_summary.tsv`, generated by `run_snps`.

   ```
   sample_name   midas_outdir
   SRR172902     /home/ubuntu/hmp_mock/midas2_output_uhgg/single_sample
   SRR172903     /home/ubuntu/hmp_mock/midas2_output_uhgg/single_sample
   ```


## Population SNVs Calling

1. **<species, sub-samples-lists> selection**

   As mentioned in the single-sample step, MIDAS 2.0 purposely hold any filter upon the single-sample pileup summary until the across-samples step.  The across-samples SNPs analysis restricts attention to "sufficiently well" covered species in "sufficiently many" samples. To be specific, only <species, sample> pair with more than 40% horizontal genome coverage (`fraction_covered`) and 5X vertical genome coverage (`mean_coverage`) would be kept. Furthermore, only "sufficiently prevalent" species with "sufficiently many" (`sample_counts`) would be included for the population SNVs analysis. Therefore, different species may have different lists of relevant samples.


2. **<site, relevant sample> selection**

   For each genomic site, a sample is considered to be "relevant" if the corresponding site read depth falls between the range defined by the input parameters `site_depth` and `site_ratio * genome_depth`; otherwise ignored for the pooled-SNPs compute.  Therefore, different genomic sites from the same species may have different panels of "relevant samples".  And genomic site prevalence can be computed as the ratio of the number of relevant samples for the given site over the total number of relevant samples for the given species.

3. **Relevant site** 

   For each species, a site is considered to be "relevant" if the site prevalence meets the range defined by the input argument `snv_type` and `site_prev`. 


4. **Population SNVs**

   (1) For each **relevant** genomic site, MIDAS 2.0 determines the set of alleles present across **all** relevant samples. Specifically, for each allele (A, C, G, T), `merge_snps` subcommand (1) counts the sample counts (sc) of **relevant samples** containing corresponding allele (`scA:scT`), and (2) sums up the read counts (rc) of the corresponding allele across all the relevant samples (`rc_G:rc_T`).

   ```
   site_id                            rc_A  rc_C  rc_G  rc_T  sc_A  sc_C  sc_G  sc_T
   gnl|Prokka|UHGG000587_14|34360|A   26    10    0     0     2     2     0     0
   gnl|Prokka|UHGG000587_11|83994|T   0     0     11    45    0     0     2     2
   ```

   (2) Across-samples major and minor allele for a single site can be computed by the metric specified via the `snps_pooled_method`: sample counts as in `prevalence` and read counts as in `abundance`. For example, the population major allele of site UHGG143484_1|905|T in the above example is T defined by accumulated read counts and C defined by accumulated sample counts. And the population minor allele refers to the second most prevalent/abundant allele.  

   (3) The sample by site matrix of the corresponding site depths and allele frequency of the above calculated across-samples minor allele for each sample will be collected and reported.

   More details about the compute can be found at [Cross-Sample SNP Analysis Tools (xsnp)](https://github.com/czbiohub/xsnp/wiki/Data-Schema-And-Computation)


### Sample commands

   ```
   midas2 merge_snps --samples_list /path/to/sample/lists --num_cores 32 /path/to/merged/midas/outdir
   ```

### Output files

- Pooled single-sample pileup summary: `snps_summary.tsv`

   ```
   sample_name  species_id  genome_length  covered_bases  total_depth  aligned_reads  mapped_reads  fraction_covered  mean_coverage
   SRR172902    100122      2560878        2108551        10782066     248700         207047        0.823             5.113
   SRR172903    100122      2560878        2300193        39263110     1180505        820736        0.898             17.069
   ```

- Per species SNPs information file: `{species_id}.snps_info.tsv.lz4`.  It contains the computed population major and minor alleles, together with all the metadata.

   ```
   site_id                             major_allele  minor_allele  sample_counts  snp_type  rc_A  rc_C  rc_G  rc_T  sc_A  sc_C  sc_G  sc_T  locus_type  gene_id           site_type  amino_acids
   gnl|Prokka|UHGG000587_14|34360|A    A             C              2             bi        26    10    0     0     2     2     0     0     CDS         UHGG000587_02083  4D         T,T,T,T
   gnl|Prokka|UHGG000587_11|83994|T    G             T              2             bi        0     0     11    45    0     0     2     2     IGR         None              None       None
   ```

- Per species site-by-sample allele frequency matrix of **population minor allele**: `{species_id}.snps_freq.tsv.lz4`.

  ```
  site_id                             SRR172902   SRR172903
  gnl|Prokka|UHGG000587_11|83994|T    0.692       0.837
  gnl|Prokka|UHGG000587_14|34360|A    0.300       0.269
  ```

- Per species site-by-sample site depth matrix: `{species_id}.snps_depth.tsv.lz4`. **Only** accounts for the alleles matching either population major or/and population minor allele.

  ```
  site_id                             SRR172902   SRR172903
  gnl|Prokka|UHGG000587_11|83994|T    13          43
  gnl|Prokka|UHGG000587_14|34360|A    10          26
  ```

## Population CNVs Profiling

MIDAS 2.0 `merge_gene` subcommands merge single-sample gene content profiling results across all the samples listed in the `samples_list`, and further on quantify each gene's presence/absence by compare the `copy_number` with the user defined minimal gene copy number to call a gene present (`min_copy`).


### Sample command

   ```
   midas2 merge_genes --samples_list /path/to/sample/lists --num_cores 8 /path/to/merged/midas/outdir
   ```

### Target output files

- Pooled single-sample gene content summary: `genes_summary.tsv`

   ```
   sample_name  species_id  pangenome_size  covered_genes  fraction_covered  mean_coverage  aligned_reads  mapped_reads  marker_depth
   SRR172902    100122      29165           2535           0.087             4.723          263395         53006         1.435
   SRR172903    100122      29165           3212           0.110             16.095         1447684        263878        10.713
   ```

- Per species gene-by-sample copy number matrix: `{species_id}.genes_copynum.tsv.lz4`

  ```
  gene_id            SRR172902     SRR172903
  UHGG000587_00401   33.969154     19.891455
  UHGG000587_01162   5.703398      2.821237
  UHGG000587_00962   2.370930      0.289325
  ```

- Per species gene-by-sample presence absence matrix: `{species_id}.genes_preabs.tsv.lz4`
 
  ```
  gene_id             SRR172902    SRR172903
  UHGG000587_00401    1            1
  UHGG000587_01162    1            1
  UHGG000587_00962    1            0 
  ```

- Per species gene-by-sample gene average coverage matrix: `{species_id}.genes_depth.tsv.lz4`

  ```
  gene_id             SRR172902   SRR172903
  UHGG000587_00401    48.747945   213.090622
  UHGG000587_01162    8.184746    30.222978
  UHGG000587_00962    3.402439    3.099448
  ```


## Across-samples Species Abundance Profile

MIDAS 2.0 also provide the subcommand `merge_species` to merge all the single-sample single copy marker genes profiling results for all the listed samples. 

### Sample command

   ```
   midas2 merge_species --samples_list /path/to/sample/lists /path/to/merged/midas/outdir
   ```

### Output files

- Pooled single-sample marker genes profiling summary: `species_prevalence.tsv`

   ```
   species_id  median_abundance  mean_abundance  median_coverage  mean_coverage  sample_counts
   102337      0.186             0.186           16.205           16.205         2
   102506      0.035             0.035           2.967            2.967          2
   ```

- Species-by-sample median marker genes coverage matrix: `species_marker_median_coverage.tsv`

   ```
   species_id   SRR172902   SRR172903
   102337       3.926       28.484
   100122       6.226       30.473
   102506       0.951       4.983
   ```

- Species-by-sample average marker genes coverage matrix: `species_marker_coverage.tsv`

   ```
   species_id   SRR172902   SRR172903
   102337      	3.926       28.484
   100122      	6.226       30.473
   102506       0.951       4.983
   ```

- Species-by-sample marker genes read counts matrix: `species_marker_read_counts.tsv`

   ```
   species_id   SRR172902   SRR172903
   102337       1565        4110
   100122       863         4410
   102506       143         734
   ```

- Species-by-sample marker genes relative abundance matrix: `species_relative_abundance.tsv`

   ```
   species_id   SRR172902   SRR172903
   102337       0.072       0.301
   100122       0.119       0.319
   102506       0.019       0.052
   ```
