
# MIDAS Reference Database

MIDAS 2.0 is a reference-based strain-level genomic variation analysis pipeline, and it also presuppose a reference database construction step has already taken place.  **MIDAS reference database (MIDASDB)** refers to a set of custom files needed for the strain-level metagenomic analysis. 

MIDAS provided a **default bacterial reference databases** (see [Figure 1](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5088602/)). [midasdb v1.2](http://lighthouse.ucsf.edu/MIDAS/midas_db_v1.2.tar.gz) were constructed from a collection of 5952 bacterial species clusters representing 31,007 high-quality bacterial genome. 

In the past few years, the number of sequenced microbial genomes have increased vastly, in particular with the addition of metagenome-assembled genomes (MAGs) sequenced from varied habitats. Therefore, it is necessary to update MIDAS reference database accordingly. On the other hand, processing the large amount of available genome sequences poses a significant compute challenge. For MIDAS 2.0, instead of generating the species clusters from scratch, we take advantage of several published collections of prokaryotic genome databases, and build MIDAS reference databases individually. 


# Database Infrastructure

To meet the challenge of increased number of available genome sequences, MIDAS 2.0 implemented a new database infrastructure, geared to run on [AWS Batch](https://aws.amazon.com/batch/) and [S3](https://aws.amazon.com/s3/), to achieve [elastic scaling](https://github.com/czbiohub/pairani/wiki) for building MIDAS reference databases. To be specific, the MIDAS reference database construction step can be executed in AWS using hundreds of r5d.24xlarge instances over a period of a couple of days, depositing built products in S3.  For example, it takes ~80k to build the species pan-genome for all 47,894 species of GTDB r202 in a week.


## Table of Content

The new database infrastructure reads in a table of contents (TOC) file, containing genome-to-species assignment and a choice of representative genome for each species cluster.  One TOC file (`genomes.tsv`) per MIDAS reference database, listing the genome-to-species assignment for all genomes, with per genome each row. The TOC file has four columns, among which `genome_is_representative` specify whether the `genome` is the representative genome for the corresponding `species`. Only one `representative` per `species`.  

```
genome              species      representative        genome_is_representative
GUT_GENOME138501    104351       GUT_GENOME269084      0
GUT_GENOME269084    104351       GUT_GENOME269084      1
```

By default, MIDAS 2.0 inherits the representative genome assignments from published prokaryotic genome databases. Inspired by the importance of selecting proper reference genome for accurate template-based SNP calling, this new infrastructure empowers user the flexibility to dynamically re-assign the representative genomes, simply by modifying the `genomes.tsv` file accordingly.


## Prokaryotic Genome Databases

As mentioned above, instead of generating species clusters from scratch, we produced MIDAS Reference Database based on published, publicly available prokaryotic genomes databases. And at this moment, there are two pre-computed MIDAS reference databases readily available for users to download, as shown in the following command:

```
$ midas2 database --list
uhgg 286997 genomes from 4644 species version 1.0
gtdb 258405 genomes from 47893 species version r202
```

### Unified Human Gastrointestinal Genome (UHGG)

A collection of 286,997 genomes assembled from metagenomes, isolates and single cells from human stool samples has been clustered into 4,644 gut-only species in [UHGG 1.0 catalogues](http://ftp.ebi.ac.uk/pub/databases/metagenomics/mgnify_genomes/human-gut/v1.0/). The collection of all the UHGG genomes were mirrored in [S3](s3://jason.shi-bucket/IGGdb2.0/clean_set/), which serves as the input to the database construction. [Six-digit numeric species ids](s3://jason.shi-bucket/IGGdb2.0/alt_species_ids.tsv) were arbitrarily assigned. Instead of species name, these `species_id` are used as species identifier in all the reports generated by MIDAS 2.0.

## Genome Taxonomy Database (GTDB)

[GTDB R06-RS202](https://gtdb.ecogenomic.org/stats/r202) contains 45,555 bacterial and 2,339 archaeal species clusters spanning 258,406 genomes, released on April 27th, 2021. The genome members for each species cluster is specified in the [sp_clusters_r202.tsv](https://data.ace.uq.edu.au/public/gtdb/data/releases/release202/202.0/auxillary_files/sp_clusters_r202.tsv), upon which order six-digit numeric species ids are assigned. GTDB only provided the sequences of the representative genomes, and we downloaded all the genomes from NCBI genomes repository using [genome_updater](https://github.com/pirovc/genome_updater). 


# Database Target Layout and Construction

MIDAS reference database (MIDASDB) primarily consist of three parts: rep-genome databases, pan-genome databases, and universal single copy genes (SGC) marker database. The target layout of any MIDASDB follow the same relative structure, based on the base directory of the database, both in the S3 bucket and locally. The following toy example demonstrates the major steps to construct the MIDASDB and the target layout using a toy collection of genomes with only one species cluster `species1` with two genomes (`genome1` and `genome2`).


![Figure 1: MIDAS Reference Database Target Layout and Construction Steps](https://github.com/czbiohub/MIDAS2.0/blob/master/docs/images/Fig.DB.Layout.png?raw=true "Title")


## Inputs

The input collection of genomes need to be organized in the format as `cleaned_genomes/<species>/<genome>.fa.lz4`. And the table of content **`genomes.tsv`** file needs to be generated accordingly, with randomly assigned six-digit `species_id`, to replace the species name. The `genome` name can be kept as it is.

```
genome     species    representative    genome_is_representative
genome1    100001     genome2           0
genome2    100001     genome2           1
```

## Rep-Genome Databases

The genome annotation for all the genomes were done by [Prokka](https://github.com/tseemann/prokka), and the annotated genes were kept under the directory of `genes_annotations/<species>/<genome>`. The rep-genome databases for the SNPs module analysis only included the gene annotations and sequences for the **representative genomes**, as specified in the TOC.

```
gene_annotations/100001/genome2/genome2.fna.lz4
gene_annotations/100001/genome2/genome2.ffn.lz4
gene_annotations/100001/genome2/genome2.genes.lz4
```

## SCG Marker Database

Marker genes are defined as universal, single-copy gene families. MIDAS uses a subset (15) of the [PhyEco gene families](https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0077033). The pre-computed HMM model of this set of 15 single copy genes (SCGs) are available at:

```
s3://microbiome-pollardlab/uhgg_v1/marker_gene_models/phyeco/marker_genes.hmm.lz4
s3://microbiome-pollardlab/uhgg_v1/marker_gene_models/phyeco/marker_genes.mapping_cutoffs.lz4
```

For each annotated genome, the homologs of 15 SCGs were identified with `hmmsearch`, as well as the mapping of gene id to corresponding marker gene id, under the directory of `marker_genes/phyeco/temp/<species>/<genome>`.

```
marker_genes/phyeco/temp/100001/genome2/genome2.markers.fa
marker_genes/phyeco/temp/100001/genome2/genome2.markers.map
```

For all the **representative genomes**, the identified marker genes were concatenated into monolithic **`marker_genes.fa`**, from which `hs-blastn` index would be constructed. The indexed `marker_genes.fa` serves as the SCG marker databases.

```
marker_genes/phyeco/marker_genes.fa
marker_genes/phyeco/marker_genes.fa.sa
marker_genes/phyeco/marker_genes.fa.bwt
marker_genes/phyeco/marker_genes.fa.sequence
```

## Pan-Genome Databases

Species-level pan-genome refers to the set of non-redundant genes that represent the genetic diversity within one species cluster. In order to construct the pan-genome database for each species, the first step if to concatenate the annotated genes from its **all genome members** into `pangenomes/100001/genes.ffn`. The second step, which is also the most time-consuming step, is to cluster the concatenated genes based on 99% percent identity (PID) using [`vsearch`](https://github.com/torognes/vsearch). Each cluster was represented by the gene at its center - centroid gene (`centroids.99.ffn`). The `centroid.99` genes were further on clustered to 95, 90, ..., PID, respectively, and the mapping relationships were listed in `centroid_info.txt`. The top level `centroids.ffn` file represents the 99 percent identity clusters, and serves as the species pan-genome databases. Reads are aligned to the pan-genome databases to determine the gene content of strains in a sample (`midas run_genes`), and reads can optionally aggregated into gene clusters at any of the lower clustering thresholds across samples (`midas2 merge_gene`).

```
pangenomes/100001/centroids.ffn
pangenomes/100001/centroid_info.txt
```
