

## Pangenome Profiling

Species-level pangenome refers to the set of non-redundant genes (centroids) clustered from all the genomes within one species cluster. Species in the restricted species profile would be concatenated and used to build the sample-specific pangenome database, to which reads were aligned using Bowtie2. Per species per centroid **copy number** were computed in three steps: (1) Per centroid, reads alignment metrics, .e.g `mapped reads` and `mean coverage`, were computed; (2) Per species, median reads coverage of all the mapped centroids corresponding to the 15 universal single-copy genes (SCGs); (3) Per centroid, `copy number` were computed and gene presence/absence were inferred.

### Sample command

   ```
   midas2 run_species --sample_name ${sample_name} -1 /path/to/R1 -2 /path/to/R2 \
         --midasdb_name uhgg --midasdb_dir /path/to/local/midasdb \
         --num_cores 8 /path/to/base/midas/output

   midas2 run_genes --sample_name ${sample_name} -1 /path/to/R1 -2 /path/to/R2 \
          --midasdb_name uhgg --midasdb_dir /path/to/local/midasdb \
          --select_by median_marker_coverage,unique_fraction_covered \
          --select_threshold=2,0.5 \
          --num_cores 12 /path/to/base/midas/output
   ```

### Output files

- Reads mapping and gene coverage summary: `genes_summary.tsv`. 

  - By default, reads aligned to more than one genomic location equally well (`MAPQ=0, 1`) are ignored.

   ```
   species_id  pangenome_size  covered_genes  fraction_covered  mean_coverage  aligned_reads  mapped_reads   marker_coverage
   102337      15578           4468           0.287             16.213         1650361        450353         20.213
   102506      731186          4733           0.006             3.803          681335         37272          2.140
   ```

- Per-species gene content profiling results: `{species_id}.genes.tsv.lz4`: 

   ```
   gene_id              gene_length  aligned_reads  mapped_reads  mean_coverage  fraction_covered  copy_number
   UHGG143901_00483     555          14             6             2.961538       0.234234          1.384035
   UHGG143901_03589     384          103            57            32.840708      0.294271          15.347667
   UHGG143902_04031     207          9              2             1.737500       0.386473          0.811997
   ```

***



# Across-samples Analysis

Population metagenomic SNVs calling and pangenome CNVs estimation can be accomplished by merging single-sample analysis results using the subcommands `merge_snps` and `merge_genes`.  All across-samples subcommands require the input file `samples_list`: a TSV file with `sample_name` and single-sample midas output directory `midas_outdir`. For example, given the following `samples_list`, `merge_snps` would expect to locate `/home/ubuntu/hmp_mock/midas2_output_uhgg/single_sample/SRR172903/snps/snps_summary.tsv`, generated by `run_snps`.

   ```
   sample_name   midas_outdir
   SRR172902     /home/ubuntu/hmp_mock/midas2_output_uhgg/single_sample
   SRR172903     /home/ubuntu/hmp_mock/midas2_output_uhgg/single_sample
   ```


## Population CNVs Profiling

MIDAS 2.0 `merge_gene` subcommands merge single-sample gene content profiling results across all the samples listed in the `samples_list`, and further on quantify each gene's presence/absence by compare the `copy_number` with the user defined minimal gene copy number to call a gene present (`min_copy`).


### Sample command

   ```
   midas2 merge_genes --samples_list /path/to/sample/lists --num_cores 8 /path/to/merged/midas/outdir
   ```

### Target output files

- Pooled single-sample gene content summary: `genes_summary.tsv`

   ```
   sample_name  species_id  pangenome_size  covered_genes  fraction_covered  mean_coverage  aligned_reads  mapped_reads  marker_depth
   SRR172902    100122      29165           2535           0.087             4.723          263395         53006         1.435
   SRR172903    100122      29165           3212           0.110             16.095         1447684        263878        10.713
   ```

- Per species gene-by-sample copy number matrix: `{species_id}.genes_copynum.tsv.lz4`

  ```
  gene_id            SRR172902     SRR172903
  UHGG000587_00401   33.969154     19.891455
  UHGG000587_01162   5.703398      2.821237
  UHGG000587_00962   2.370930      0.289325
  ```

- Per species gene-by-sample presence absence matrix: `{species_id}.genes_preabs.tsv.lz4`
 
  ```
  gene_id             SRR172902    SRR172903
  UHGG000587_00401    1            1
  UHGG000587_01162    1            1
  UHGG000587_00962    1            0 
  ```

- Per species gene-by-sample gene average coverage matrix: `{species_id}.genes_depth.tsv.lz4`

  ```
  gene_id             SRR172902   SRR172903
  UHGG000587_00401    48.747945   213.090622
  UHGG000587_01162    8.184746    30.222978
  UHGG000587_00962    3.402439    3.099448
  ```


## Across-samples Species Abundance Profile

MIDAS 2.0 also provide the subcommand `merge_species` to merge all the single-sample single copy marker genes profiling results for all the listed samples. 

### Sample command

   ```
   midas2 merge_species --samples_list /path/to/sample/lists /path/to/merged/midas/outdir
   ```

### Output files

- Pooled single-sample marker genes profiling summary: `species_prevalence.tsv`

   ```
   species_id  median_abundance  mean_abundance  median_coverage  mean_coverage  sample_counts
   102337      0.186             0.186           16.205           16.205         2
   102506      0.035             0.035           2.967            2.967          2
   ```

- Species-by-sample median marker genes coverage matrix: `species_marker_median_coverage.tsv`

   ```
   species_id   SRR172902   SRR172903
   102337       3.926       28.484
   100122       6.226       30.473
   102506       0.951       4.983
   ```

- Species-by-sample average marker genes coverage matrix: `species_marker_coverage.tsv`

   ```
   species_id   SRR172902   SRR172903
   102337      	3.926       28.484
   100122      	6.226       30.473
   102506       0.951       4.983
   ```

- Species-by-sample marker genes read counts matrix: `species_marker_read_counts.tsv`

   ```
   species_id   SRR172902   SRR172903
   102337       1565        4110
   100122       863         4410
   102506       143         734
   ```

- Species-by-sample marker genes relative abundance matrix: `species_relative_abundance.tsv`

   ```
   species_id   SRR172902   SRR172903
   102337       0.072       0.301
   100122       0.119       0.319
   102506       0.019       0.052
   ```
