# Microbiome - Integrated Gut Genome Tools

A collection of 286,997 metagenome-assembled genomes (MAGs) from human gut samples has been clustered into 4,644 species in an effort similar to [IGGdb 1.0](https://github.com/snayfach/IGGdb).   We refer to this new collection as IGGdb 2.0, IGG 2.0, IGG+, or sometimes simply IGG.

This repository contains tools for building a [MIDAS](https://github.com/snayfach/MIDAS) database from IGG 2.0, geared to run on [AWS Batch](https://aws.amazon.com/batch/) as described in [PairANI](https://github.com/czbiohub/pairani/wiki).  The resulting database will be located at [s3://microbiome-igg/2.0](http://microbiome-igg.s3.amazonaws.com/2.0/README.TXT).

Our project to update [MIDAS for IGGdb 1.0](https://github.com/czbiohub/MIDAS-IGGdb/blob/master/README.md) is on hold.  We will first update MIDAS to run with this new database, directly from S3.

# Inputs

The IGG 2.0 genomes, genome-to-species assignments, and a choice of representative genome for each species, were provided by [Alexandre Almeida](https://www.ebi.ac.uk/about/people/alexandre-almeida) of EBI and mirrored in S3 by [Jason Shi](http://docpollard.org/people/jason-shi/), whose `s3://jason.shi-bucket/IGGdb2.0/clean_set/` serves as input to the tools in this repository.

Numeric species ids were arbitrarily assigned by Jason Shi in `s3://jason.shi-bucket/IGGdb2.0/alt_species_ids.tsv`.

# Target layout in S3

A table of contents listing all MAGs will be located at [s3://microbiome-igg/2.0/genomes.tsv](http://microbiome-igg.s3.amazonaws.com/2.0/genomes.tsv).  It will look like
```
genome              species      representative        genome_is_representative
GUT_GENOME138501    104351       GUT_GENOME269084      0
GUT_GENOME269084    104351       GUT_GENOME269084      1
...
```
with 4644 rows below the headers.


Every species has a single representative genome, which could be used to identify the species, if we didn't have the purely numeric 6-digit species ids.  

Column `genome_is_representative` is defined as `genome == representative`.

## Genomes

```
s3://igg/2.0/genomes/GUT_GENOME138501/GUT_GENOME138501.{fna, ffn, faa, tsv, err, log, ...}.lz4
```

The fna file contains a metagenome-assembled genome (MAG) from IGG+.  The rest are generated by prokka.

## Pangenomes
```
s3://igg/2.0/pangenomes/104351/104351_{genes.ffn, centroids.ffn, gene_info.txt}.lz4
s3://igg/2.0/pangenomes/104351/temp/104351_centroids.{99, 95, 90, 85, 80, 75}.ffn.lz4
s3://igg/2.0/pangenomes/104351/temp/104351_uclust.{99, 95, 90, 85, 80, 75}.txt.lz4
```
The `{YYYYYY}_genes.ffn` for a species is a concatenation of all `GUT_GENOME{XXXXXX}.ffn` produced by prokka for the genomes in that species, with some minor cleanup: 

  * extra newlines have been removed from the gene sequences, so that each gene sequence occupies a single line below the gene header

  * DNA letters have been converted to uppercase, and 

  * degenerate genes (with empty sequences or headers containing "|") have been excluded

The temp files are produced by clustering `genes.ffn` with `vsearch` to 99, 95, ... percent identity, respectively.

The top level `centroids.ffn` file represents the 99 percent identity clusters -- with each cluster represented by the gene at its center.

For every gene from `genes.ffn`, the centroids of the clusters containing that gene are listed in `gene_info.txt`.  For example, in the following excerpt, we see 5 genes that belong to two different 99-pid clusters, but only to a single 95-pid cluster centered around the gene `GUT_GENOME032486_01058`.
```
gene_id                   centroid_99               centroid_95
GUT_GENOME032486_01058    GUT_GENOME032486_01058    GUT_GENOME032486_01058
GUT_GENOME036826_01544    GUT_GENOME036826_01544    GUT_GENOME032486_01058
GUT_GENOME122726_01472    GUT_GENOME036826_01544    GUT_GENOME032486_01058
GUT_GENOME125418_01744    GUT_GENOME036826_01544    GUT_GENOME032486_01058
GUT_GENOME153817_01616    GUT_GENOME036826_01544    GUT_GENOME032486_01058
```