# Integrated Gut Genome Tools

A collection of 286,997 genomes assembled from metagenomes, isolates and single cells from human stool samples has been clustered into 4,644 species in an effort similar to [IGGdb 1.0](https://github.com/snayfach/IGGdb).   We refer to this new collection as IGGdb 2.0, IGG 2.0, IGG+, or simply IGG.  Perhaps the most important difference with respect to the original IGGdb 1.0 is that this new collection contains only gut genomes.

This repository contains tools for building a [MIDAS](https://github.com/snayfach/MIDAS) database from IGG 2.0, geared to run on [AWS Batch](https://aws.amazon.com/batch/) as described in [PairANI](https://github.com/czbiohub/pairani/wiki).  The resulting database will be located at [s3://microbiome-igg/2.0](http://microbiome-igg.s3.amazonaws.com/2.0/README.TXT).

Our project to update [MIDAS for IGGdb 1.0](https://github.com/czbiohub/MIDAS-IGGdb/blob/master/README.md) is on hold.  We will first update MIDAS to run with this new database, directly from S3.

# Inputs

The IGG 2.0 genomes, genome-to-species assignments, and a choice of representative genome for each species, were provided by [Alexandre Almeida](https://www.ebi.ac.uk/about/people/alexandre-almeida) of EBI and mirrored in S3 by [Jason Shi](http://docpollard.org/people/jason-shi/), whose `s3://jason.shi-bucket/IGGdb2.0/clean_set/` serves as input to the tools in this repository.

Numeric species ids were arbitrarily assigned by Jason Shi in `s3://jason.shi-bucket/IGGdb2.0/alt_species_ids.tsv`.

Question for Alexandre:   What were the criteria for clustering genomes into species, and for choosing a representative genome for each species?

# Target Layout in S3

A table of contents listing all genomes will be located at [s3://microbiome-igg/2.0/genomes.tsv](http://microbiome-igg.s3.amazonaws.com/2.0/genomes.tsv).  It will look like
```
genome              species      representative        genome_is_representative
GUT_GENOME138501    104351       GUT_GENOME269084      0
GUT_GENOME269084    104351       GUT_GENOME269084      1
...
```
with 286,997 rows below the headers (one per genome) and 4,644 distinct species.

Every species has a unique representative genome.

Column `genome_is_representative` is defined as `genome == representative`.

## Gene Annotations

```
s3://microbiome-igg/2.0/prodigal/GUT_GENOME138501.{fna, faa, gff, log}.lz4
```

The ORFs are generated by Prodigal. The genome fna file need to be retrieved from `clean_set`. 

## Pan-Genomes
```
s3://microbiome-igg/2.0/species/GUT_GENOME138501/{genes.ffn, centroids.ffn, gene_info.txt}
s3://microbiome-igg/2.0/species/GUT_GENOME138501/temp/centroids.{99, 95, 90, 85, 80, 75}.ffn
s3://microbiome-igg/2.0/species/GUT_GENOME138501/temp/uclust.{99, 95, 90, 85, 80, 75}.txt
```
The species pangenome `{YYYYYY}/genes.ffn` is a concatenation of all `GUT_GENOME{XXXXXX}.fna` produced by Prodigal for the genomes in that species, with some minor cleanup: 

  * extra newlines have been removed from the gene sequences, so that each gene sequence occupies a single line below the gene header

  * DNA letters have been converted to uppercase, and 

  * degenerate genes (with empty sequences or headers containing "|") have been excluded

The temp files are produced by clustering `genes.ffn` with `vsearch` to 99, 95, ... percent identity (PID), respectively.

The top level `centroids.ffn` file represents the 99 percent identity clusters -- with each cluster represented by the gene at its center.

For every gene from `genes.ffn`, the centroids of the clusters containing that gene are listed in `gene_info.txt`.  For example, in the following excerpt, we see 5 genes that belong to two different 99-PID clusters, but only to a single 95-PID cluster centered around the gene `GUT_GENOME032486_01058`.
```
gene_id                   centroid_99               centroid_95
GUT_GENOME032486_01058    GUT_GENOME032486_01058    GUT_GENOME032486_01058
GUT_GENOME036826_01544    GUT_GENOME036826_01544    GUT_GENOME032486_01058
GUT_GENOME122726_01472    GUT_GENOME036826_01544    GUT_GENOME032486_01058
GUT_GENOME125418_01744    GUT_GENOME036826_01544    GUT_GENOME032486_01058
GUT_GENOME153817_01616    GUT_GENOME036826_01544    GUT_GENOME032486_01058
```

## Marker Genes
```
s3://microbiome-igg/2.0/marker_genes/phyeco.fa
s3://microbiome-igg/2.0/marker_genes/phyeco.map
s3://microbiome-igg/2.0/marker_genes/phyeco.fa.{sa, bwt, sequence}
```
The `phyeco.fa.{sa, bwt, sequence}` is the HS-BLASTN index for the `phyeco.fa`.

In order to generate the target files in parallel, we ran HMM search against the protein sequences of the ORFs for each genome:

```
s3://microbiome-igg/2.0/marker_genes/temp/GUT_GENOME138501.hmmsearch
s3://microbiome-igg/2.0/marker_genes/temp/GUT_GENOME138501.phyeco.fa
s3://microbiome-igg/2.0/marker_genes/temp/GUT_GENOME138501.phyeco.map
```

The `GUT_GENOME{XXXXXX}.hmmsearch` is the HMMER results of searching reads aligned the 15 phyeco genes. 
The `GUT_GENOME{XXXXXX}.phyeco.fa` is the FASTA file for the aligned phyeco genes, and `GUT_GENOME{XXXXXX}.phyeco.map` is needed for building the marker genes database.

The precomputed HMM model for the 15 phyeco genes and gene-specific cutoffs are available at:

```
s3://microbiome-igg/2.0/phyeco.hmm
s3://microbiome-igg/2.0/phyeco.mapping_cutoffs
```

After running HMM and extract the phyeco genes for each genomes, concatenate the FA files for all the representative genomes into `phyeco.fa`; concatenate the phyeco mapping files for all the genomes into `phyeco.map`.

```
## concatenate the FASTA sequences
awk '$3 == 1 {print $1 }' mapfile  | xargs -Ixx -P 32 bash -c "cp marker_genes/temp/xx.phyeco.fa marker_genes/phyeco/xx.phyeco.fa"
cat marker_genes/phyeco/xx.phyeco.fa > marker_genes/phyeco.fa

## concatenate the MAP files
grep -v "alt" alt_species_ids.tsv | cut -f2 |  xargs -Irep bash -c "awk -v pat=rep '\$2 == pat {print}' mapfile | cut -f1 | xargs -Igenome bash -c 'cat temp/genome.phyeco.map' > phyeco/rep.phyeco.map "
cat marker_genes/phyeco/*.phyeco.map > marker_genes/phyeco.map

## Add header to MAP files
echo -e  "species_id\tgenome_id\tgene_id\gene_length\tmarker_id" | cat -  phyeco.map > phyeco.temp && mv phyeco.temp phyeco.map

## Make hs-blastn index
hs-blastn index marker_genes/phyeco.fa
```

## Hacks/*.sh

Individual bash scripts to call Progidal and generate HMM marker genes database. 


# See Also

[Operations](https://github.com/czbiohub/iggtools/wiki/Operations)
***

***

***

***
